### PR TITLE
Lab2 - Implement bubble and insertion sort, and tests

### LINKED ISSUE #414: Build lab2 code, implement bubble sorting and insertion sorting

### PR DESCRIPTION
**Pull request by alfredxiao**
Fixes #414
Review colleague's PR #293 

Changes proposed in this PR:
- bubble sort
- insertion sort

### CODE REVIEW

File: 02_bubble/alfredxiao/main.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func makeCopy(ns []int) []int {
+	return append([]int{}, ns...)
```
Comment (n0npax) [inline 02_bubble/alfredxiao/main.go:12]: 👍

File: 02_bubble/alfredxiao/main.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	n := len(copy)
+	for i := n - 1; i > 0; i-- {
+		for j := 0; j < i; j++ {
+			maybeSwap(copy, j, j+1)
+		}
+	}
+	return copy
+}
+
+func maybeSwap(ns []int, i, j int) bool {
```
Comment (n0npax) [inline 02_bubble/alfredxiao/main.go:26]: name is ugly. I'd prefer sth like `swapIfLower`

File: 02_bubble/alfredxiao/main.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	if ns[i] > ns[j] {
+		ns[i], ns[j] = ns[j], ns[i]
+		return true
+	}
+	return false
+}
+
+func insertion(ns []int) []int {
```
Comment (n0npax) [inline 02_bubble/alfredxiao/main.go:34]: you forgot to copy input slice. `s := makeCopy(ns)`

File: 02_bubble/alfredxiao/main_test.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	assert.Equal(t, []int{3, 2, 1}, ns)
+}
+
+func TestInsertionEdgeCases(t *testing.T) {
+	assert.Equal(t, []int{}, insertion([]int{}))
+	assert.Equal(t, []int{1}, insertion([]int{1}))
+	assert.Equal(t, []int{1, 2}, insertion([]int{2, 1}))
+}
+
+func TestInsertionBestCase(t *testing.T) {
```
Comment (n0npax) [inline 02_bubble/alfredxiao/main_test.go]: This test is the same as `TestBubbleBestCase`

I'd suggest passing function as an argument to test: i.e
```
algo := {"bubble": bubble, "insertion": insertion}
for _,f := range algo {
    assert.Equal(t, []int{2, 3, 3, 6, 6, 8, 9}, f([]int{3, 2, 3, 6, 8, 6, 9}))
}
```
as next step you can also keep slices with data and pass to the test:
```
algo := {"bubble": bubble, "insertion": insertion}
slices := [{"input":[]int{3,1,2}, "expected":[]int{1,2,3} }, ..., ..., ... ]
for _,f := range algo {
    for _, s := range slices {
      assert.Equal(t, s.expected, f(s.input))
  }
}
```
as next improvement you can put assert inside subtest
```
algo := {"bubble": bubble, "insertion": insertion}
slices := [{"input":[]int{3,1,2}, "expected":[]int{1,2,3} }, ..., ..., ... ]
for a, f := range algo {
    for _, s := range slices {
      t.Run(a, func(t *testing.T) {
      assert.Equal(t, s.expected, f(s.input))
    }
  }
}
```

**Review by n0npax — CHANGES_REQUESTED** [review]

File: 02_bubble/alfredxiao/main.go (hunk 3 above)
Comment (alfredxiao) [reply 02_bubble/alfredxiao/main.go:34]: good catch

File: 02_bubble/alfredxiao/main.go (hunk 2 above)
Comment (alfredxiao) [reply 02_bubble/alfredxiao/main.go:26]: I have found some people are not comfortable with names in the `maybeVerb` form. But this is actually commonly used. If you search in JDK source code, you can find many occurrences of maybe* there. Although I agree swapIfLower is a good candidate, naming is always a very difficult and subjective thing. You can actually find mongodb uses maybeSwap in https://github.com/mongodb/mongo/search?q=maybeSwap&unscoped_q=maybeSwap

File: 02_bubble/alfredxiao/main_test.go (hunk 4 above)
Comment (alfredxiao) [reply 02_bubble/alfredxiao/main_test.go]: good suggestion, I have applied some refactoring there

Comment (n0npax) [reply 02_bubble/alfredxiao/main_test.go]: I guess you forgot to push. `628a2e8deda6e0152b18ae247b4b6b5e194cdc69` doesn't touch this file.

File: 02_bubble/alfredxiao/main.go (hunk 2 above)
Comment (n0npax) [reply 02_bubble/alfredxiao/main.go:26]: Clean code says function name should be descriptive and blah blah blah (holly war, let's skip this :) ).
Don't like this but it's not a blocker in my mind.

**Review by n0npax — DISMISSED** [review]

File: 02_bubble/alfredxiao/main_test.go (hunk 4 above)
Comment (alfredxiao) [reply 02_bubble/alfredxiao/main_test.go]: pardon me, pushed now

**Review by n0npax — APPROVED** [review]

File: 02_bubble/alfredxiao/main.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+func makeCopy(ns []int) []int {
+	return append([]int{}, ns...)
+}
+
+func bubble(ns []int) []int {
+	copy := makeCopy(ns)
+	n := len(copy)
+	for i := n - 1; i > 0; i-- {
+		for j := 0; j < i; j++ {
+			maybeSwap(copy, j, j+1)
```
Comment (juliaogris) [inline 02_bubble/alfredxiao/main.go:20]: I agree with @n0npax - awkward name, does more than you want here (ignored returned value).
IMO it'd be more readable if you just inlined :
```go
if copy[j] > copy[j+1] {
    copy[j], copy[j+1] = copy[j+1], copy[j]
}
```
Also: `copy` is a builtin identifier that you are shadowing here.

**Review by juliaogris — APPROVED** [review]
LGTM.
However, I too would have avoided the `maybeSwap` function. It is a question of taste I suppose.

**Pull request merged**